### ISSUE TITLE
The Feature Request of Loading Quantized TorchScript Model on Windows with libtorch

### ISSUE THREAD

**Issue by KaiyuYue** [pytorch/pytorch#31684]
## 🚀 Feature Request [Solved] [Wait for Future Release Support]

The `torch::jit::load` works well for loading both quantized model and un-quantized model on Linux and MacOS. But it will throw an error on Windows when loading the quantized model.

## :cookie: Currently, official FBGEMM can be compiled with PyTorch for Windows.

PR:  #33250 

## :cookie:  ~~Current Unofficial Method~~

Currently we compile `libtorch-1.3.1` with [FBGEMM marian-nmt version](https://github.com/marian-nmt/FBGEMM) by @ykim362 on windows successfully. And the quantized model works pretty well on windows (200 ms -> 57 ms). Thanks for the effort from @ykim362. Hope `fbgemm` for windows can be aggregated in the **future/next** release of PyTorch. You know, the need of quantization for windows is as much as for Linux and ARM.

Some details:
- PyTorch-1.3.1 : [released branch with tag-1.3.1](https://github.com/pytorch/pytorch/tree/v1.3.1)
- FBGEMM : [marian-nmt version](https://github.com/marian-nmt/FBGEMM) by @ykim362 
- `set CAFFE2=OFF` when compiling on windows 

## To Reproduce

Quick steps to reproduce the behavior:

1. Use a model under [torchvision/models/quantization](https://github.com/pytorch/vision/tree/master/torchvision/models/quantization), for example, mobilenet.
2. In model definition py file, add

```python
if __name__ == '__main__':
    model = mobilenet_v2(pretrained=False, progress=True, quantize=True)
    torch.jit.save(torch.jit.script(model), 'mov2.pt')
```
3. Load the scripted model in Linux, MacOS, and Windows

```cpp
#include <torch/torch.h>
#include <torch/script.h> 
#include <iostream>
#include <memory>

int main(int argc, const char* argv[]) {
    torch::jit::script::Module quant_model;
    quant_model = torch::jit::load(argv[1]);
}
```

## Expected behavior

The `torch::jit::load` works well for loading both quantized model and un-quantized model on Linux and MacOS. But it will throw an error on Windows when loading the quantized model:

```bash
The field 'training' was left unitialized after __setstate__, but expected a value of type 'bool' (postSetStateValidate at ..\..\torch\csrc\jit\import.cpp:52)
(no backtrace available)
```

## Environment

 - PyTorch Version: 1.3.1
- LibTorch Version: 1.3.1
 - OS: Ubuntu, MacOS, and Windows10 

cc @suo @jerryzh168 @jianyuh @dzhulgakov @raghuramank100 @jamesr66a @soumith

**Comment by KaiyuYue** [comment]
I try to use jit.load with python and know that `RuntimeError: Didn't find engine for operation quantized::conv2d_prepack NoQEngine`. So are there any future plans to support?

```bash
Python 3.7.5 (default, Oct 31 2019, 15:18:51) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'1.4.0.dev20191229'
>>> m = torch.jit.load("epoch_q_00001.pth")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\kaiyuyue\Miniconda3\lib\site-packages\torch\jit\__init__.py", line 232, in load
    cpp_module = torch._C.import_ir_module(cu, f, map_location, _extra_files)
RuntimeError: Didn't find engine for operation quantized::conv2d_prepack NoQEngine
The above operation failed in interpreter.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/quantized/modules/conv.py", line 133
        self.groups = state[8]
        self.padding_mode = state[9]
        self.set_weight_bias(state[10], state[11])
        ~~~~~~~~~~~~~~~~~~~~ <--- HERE
        self.scale = state[12]
        self.zero_point = state[13]
Serialized   File "code/__torch__/torch/nn/quantized/modules/conv.py", line 54
    self.groups = (state)[8]
    self.padding_mode = (state)[9]
    _18 = (self).set_weight_bias((state)[10], (state)[11], )
           ~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    self.scale = (state)[12]
    self.zero_point = (state)[13]
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/quantized/modules/conv.py", line 194
    def set_weight_bias(self, w, b):
        # type: (torch.Tensor, Optional[torch.Tensor]) -> None
        self._packed_params = torch.ops.quantized.conv2d_prepack(
                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
            w, b, self.stride, self.padding, self.dilation, self.groups)
Serialized   File "code/__torch__/torch/nn/quantized/modules/conv.py", line 74, in set_weight_bias
    _30 = [_28, _29]
    _31, _32, = _23
    _33 = ops.quantized.conv2d_prepack(w, b, _27, _30, [_31, _32], _24)
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    self._packed_params = _33
    return None
```

**Comment by jerryzh168** [comment]
Can we run fbgemm in windows? cc @dskhudia, if so, could you set qengine to fbgemm and try again?

**Comment by jerryzh168** [comment]
this is how you can see the supported qengine and set qengines: https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L1426

**Comment by KaiyuYue** [comment]
> Can we run fbgemm in windows? cc @dskhudia, if so, could you set qengine to fbgemm and try again?

Hi @jerryzh168 , thanks for tracking on this issue. I have noticed there are some PRs for MSVC on windows, like [issue-360727595](https://github.com/pytorch/FBGEMM/pull/240#issue-360727595). Can this be used within `libtorch` right now? cc @dskhudia

**Comment by KaiyuYue** [comment]
Hi, @jerryzh168 @dskhudia @soumith

**Comment by KaiyuYue** [comment]
Hi @jerryzh168 @dskhudia,

Good news: Currently we compile `libtorch-1.3.1` [master code within tag-1.3.1](https://github.com/pytorch/pytorch/tree/v1.3.1) with [FBGEMM marian-nmt version](https://github.com/marian-nmt/FBGEMM) by @ykim362 on windows successfully. And the quantized model works pretty well on windows (200 ms -> 57ms). Thanks the effort from @ykim362, hope `fbgemm` for windows can be aggregated in the **future/next** release of PyTorch. You know, the need of quantization for windows is as much as for Linux and ARM.

**Comment by peterjc123** [comment]
I believe the master branch of FBGEMM is also ready to be used. Could you please confirm? cc @dskhudia

**Comment by KaiyuYue** [comment]
> I believe the master branch of FBGEMM is also ready to be used. Could you please confirm? cc @dskhudia

That’s great. Sure. I will give feedback ASAP.

**Comment by KaiyuYue** [comment]
> I believe the master branch of FBGEMM is also ready to be used. Could you please confirm? 

@dskhudia @peterjc123  Yes. It can be compiled successfully on Windows now. It works. Thanks for the effort.

**Comment by KaiyuYue** [comment]
Hi @peterjc123,
Do you have plans to support to compile for win32 ?

**Comment by peterjc123** [comment]
We are doing that through https://github.com/pytorch/pytorch/pull/33250.

**Comment by jianyuh** [comment]
#33250 has been merged in PyTorch. FBGEMM side should be unblocked for Windows. cc @shz0116, @jspark1105 , @dskhudia , @peterjc123 , @KaiyuYue

**Comment by KaiyuYue** [comment]
@jianyuh Sorry for this additional question. Are there any plans to support FBGEMM on `x86 32bit` of Windows? https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt#L401

**Comment by shz0116** [comment]
@KaiyuYue I can give it a try. But wondering how popular of using 32 bits vs. 64 bits.

**Comment by KaiyuYue** [comment]
Thanks for trying that. Yes, 64 bits is the future. But massive conventional business companies have their own software system to offer the services for society. But their software systems are commonly large and be running for a long time. So the update cycle for system including both software and hardware is also long (more than one-two years). Most of them still use old system frameworks. To support quantization tech on 32 bits, i think, is still a big requirement in industry.

**Comment by peterjc123** [comment]
As for PyTorch, we support building 32-bit binaries but we won't provide them, at least for desktop OS. According to the posts here, the number of the users that request 32-bit support for those platforms is limited. What's more, some dependencies dropped their support for x86, like CUDA.

**Comment by KaiyuYue** [comment]
Yes, agree. I think the most meaningful part of supporting on x86 is to make Quantization engine (FBGEMM)  be running with CPU for inference. Training or doing inference with GPU using x86 indeed seems like unnecessary.

**Comment by peterjc123** [comment]
Closing it because the original issue should be resolved. If users want 32-bit quantization support, they can open a new issue.

**Comment by KaiyuYue** [comment]
Thanks very much.
